src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,1048 @@
+      textureSizePaddingValueInBlocks,
+      format,
+      initMethod,
+      checkMethod,
+    } = t.params;
+
+    const info = kSizedTextureFormatInfo[format];
+
+    const origin = { x: info.blockWidth, y: info.blockHeight, z: 1 };
+    const copySize = { width: 2 * info.blockWidth, height: 2 * info.blockHeight, depth: 2 };
+    const textureSize: [number, number, number] = [3 * info.blockWidth, 3 * info.blockHeight, 3];
+
+    switch (coordinateToTest) {
+      case 'width': {
+        origin.x = originValueInBlocks * info.blockWidth;
+        copySize.width = copySizeValueInBlocks * info.blockWidth;
+        textureSize[0] =
+          origin.x + copySize.width + textureSizePaddingValueInBlocks * info.blockWidth;
+        break;
+      }
+      case 'height': {
+        origin.y = originValueInBlocks * info.blockHeight;
+        copySize.height = copySizeValueInBlocks * info.blockHeight;
+        textureSize[1] =
+          origin.y + copySize.height + textureSizePaddingValueInBlocks * info.blockHeight;
+        break;
+      }
+      case 'depth': {
+        origin.z = originValueInBlocks;
+        copySize.depth = copySizeValueInBlocks;
+        textureSize[2] = origin.z + copySize.depth + textureSizePaddingValueInBlocks;
+        break;
+      }
+    }
+
+    const rowsPerImage = copySize.height;
+    const bytesPerRow = align(copySize.width, 256);
+    const dataSize = t.requiredBytesInCopy(
+      { offset: 0, bytesPerRow, rowsPerImage },
+      format,
+      copySize
+    );
+
+    /** For testing width: we copy a (_ x 2 x 2) (in texel blocks) part of a (_ x 3 x 3) (in texel blocks)
+     * texture with origin (_, 1, 1) (in texel blocks). Similarly for other coordinates. */
+    t.uploadTextureAndVerifyCopy({
+      textureDataLayout: { offset: 0, bytesPerRow, rowsPerImage },
+      copySize,
+      dataSize,
+      origin,
+      textureSize,
+      format,
+      initMethod,
+      checkMethod,
+      changeBeforePass: 'arrays',
+    });
+  });
+
+/** Generates textureSizes which correspond to the same physicalSizeAtMipLevel
+ * including virtual sizes at mip level different from the physical ones. */
+function* textureSizeExpander({
+  format,
+  mipLevel,
+  _texturePhysicalSizeAtMipLevelInBlocks,
+}: {
+  format: SizedTextureFormat;
+  mipLevel: number;
+  _texturePhysicalSizeAtMipLevelInBlocks: GPUExtent3DDict;
+}): Generator<{ textureSize: [number, number, number] }> {
+  const info = kSizedTextureFormatInfo[format];
+
+  const widthAtThisLevel = _texturePhysicalSizeAtMipLevelInBlocks.width * info.blockWidth;
+  const heightAtThisLevel = _texturePhysicalSizeAtMipLevelInBlocks.height * info.blockHeight;
+  const textureSize: [number, number, number] = [
+    widthAtThisLevel << mipLevel,
+    heightAtThisLevel << mipLevel,
+    _texturePhysicalSizeAtMipLevelInBlocks.depth,
+  ];
+  yield {
+    textureSize,
+  };
+
+  /** We choose width and height of the texture so that the values are divisible by blockWidth and
```

Comment:
Also this doc comment is attached to the `const widthAtPrevLevel`, should just be a regular comment so it doesn't do that